### PR TITLE
Add execution phase

### PR DESCRIPTION
**Pull request by rogercloud**
1. Add git action to guard PR
2. Add execution phase

### CODE REVIEW

File: epaxos/src/execute.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+        let mut sort_vec = Vec::with_capacity(each_scc.len());
+        for (id, ins) in ins_vec.enumerate() {
+            let ins_read = ins.get_instance_read().await;
+            let ins_read = ins_read.as_ref().unwrap();
+            sort_vec.push((id, (ins_read.id.replica, ins_read.seq)));
+        }
+
+        sort_vec.sort_by(|a, b| {
+            match a.1 .1.partial_cmp(&b.1 .1) {
```
Comment (pzheng1025) [inline epaxos/src/execute.rs]: Is the extra space between `a.1` and `1` intended?

Comment (rogercloud) [reply epaxos/src/execute.rs]: this space should be deleted

File: epaxos/src/types.rs [hunk 2]
```diff
@@ -174,17 +417,23 @@ where
...
         for c in cmds {
-            let new_inst = match self.conflict[*self.id].get(c.key()) {
+            let new_inst = match self.conflict[**replica].get(c.key()) {
                 None => Some(new_inst.clone()),
                 Some(ins) => {
-                    if ins.read().await.id < new_inst.read().await.id {
-                        Some(new_inst.clone())
-                    } else {
+                    let ins = ins.get_instance_read().await;
+                    if ins.is_none()
```
Comment (pzheng1025) [inline epaxos/src/types.rs]: This logic is incorrect.

File: epaxos/src/execute.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+) -> (Vec<Vec<NodeIndex>>, DiGraph<SharedInstance<C>, ()>)
+where
+    C: Command + Clone + Send + Sync + 'static,
+{
+    let mut queue = VecDeque::new();
+    queue.push_back(ins);
+
+    let mut map = HashMap::<SharedInstance<C>, NodeIndex>::new();
+    let mut g = DiGraph::<SharedInstance<C>, ()>::new();
+    loop {
```
Comment (pzheng1025) [inline epaxos/src/execute.rs]: push deps into queue.

File: epaxos/src/execute.rs (hunk 1 above)
Comment (rogercloud) [reply epaxos/src/execute.rs]: it's the result of the formatter. Let's keep it.

File: epaxos/src/execute.rs (hunk 3 above)
Comment (rogercloud) [reply epaxos/src/execute.rs]: Also check the instance status

File: epaxos/src/execute.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    /// - the instance's status is NOT COMMITTED and NOT EXECUTED.
+    /// - the instance is empty.
+    ///
+    /// The return value is None if there's no instance to execute.
+    /// The return value is Some(Vec<...>), which is the scc vec, if there are instances to execute.
+    async fn build_scc(&mut self) -> Option<Vec<Vec<NodeIndex>>>
+    where
+        C: Command + Clone + Send + Sync + 'static,
+    {
+        // the start_ins is at leat in the stage of COMMITTED
```
Comment (pzheng1025) [inline epaxos/src/execute.rs]: typo: at least

**Review by pzheng1025 — APPROVED** [review]

**Review by pzheng1025 — APPROVED** [review]

**Pull request merged**